sstables/checksummed_data_source.cc

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <cstdint>
+
+#include <seastar/core/bitops.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+#include "types.hh"
+#include "exceptions.hh"
+#include "checksum_utils.hh"
+
+namespace sstables {
+
+extern logging::logger sstlog;
+
+// File data source implementation for SSTables with attached checksum
+// data and no compression
+template <ChecksumUtils ChecksumType>
+class checksummed_file_data_source_impl : public data_source_impl {
+    std::optional<input_stream<char>> _input_stream;
```

Comment:
I think we can get away without the optional. I think a default-constructed `input_stream` also has a valid `close()` method, no? Avoiding calling close seems to be the only reason to make this an optional.

Reply:
It does, but the underlying data source doesn't. It would attempt to close the data source implementation which is a non-initialized unique pointer.

Reply:
I see, I think we should patch it so it checks the pointer first. As a workaround, we can keep using an optional here.

Reply:
> It does, but the underlying data source doesn't. It would attempt to close the data source implementation which is a non-initialized unique pointer.

I don't understand why this happens.

Reply:
The code goes like this:
`checksummed_file_data_source_impl::close()` -> `input_stream::close()` -> `data_source::close()` -> `data_source_impl::close()`

The `input_stream` is currently an optional data member that the constructor won't create if there is nothing to read. This is a minor optimization I borrowed from the compressed data source impl and I didn't have in the first iterations.

If the `input_stream` were default constructed instead of optional, and we tried to close it, it would attempt to close the underlying data source. This would cause a null pointer dereference on the underlying data source impl here: https://github.com/scylladb/seastar/blob/d28c9a25c7e8f4e5dc302d73e5040980efa99beb/include/seastar/core/iostream.hh#L96
The default constructor for the data source was introduced in https://github.com/scylladb/seastar/commit/11f5a8e3982d4036d44665156da3632172e94906.

Reply:
Ah, I wasn't aware you can default-construct an input_stream, that's a bug.